Error in MoxyReflector class in project with several modules

The kotlon-project has the following structure:
- App 1 (with Moxy)
- App 2 (with Moxy)
- Library 1 (with Moxy)
- Library 2
- Library 3.
Library 3 is available from the Library 2, and Library 2 from Library 1. Library 1 must be available from both applications. To do this, in the "default Config {}" section of the file "build.gradle " Libraries 1 added instruction:
        kapt {
            arguments {
                arg("moxyReflectorPackage", "top.inttech.app.employment_service")
            }
        }
Annotation "@RegisterMoxyReflectorPackages("top.inttech.app.employment_service")" added to App class of application.
When you try to build the file, you receive an error message:
..........\MoxyReflector.java:34: error: cannot find symbol
		return viewStateProvider.getViewState();
		       ^
  symbol:   variable viewStateProvider
  location: class MoxyReflector
...........\MoxyReflector.java:38: error: incompatible types: Map<Class<?>,Object> cannot be converted to Map<Class<?>,List<Object>>
		return sViewStateProviders;
		       ^
2 errors
How to avoid this mistake?

--- a/moxy-compiler/src/main/java/com/arellomobile/mvp/compiler/reflector/MoxyReflectorGenerator.java
+++ b/moxy-compiler/src/main/java/com/arellomobile/mvp/compiler/reflector/MoxyReflectorGenerator.java
@@ -90,13 +90,13 @@ public class MoxyReflectorGenerator {
 			classBuilder.addMethod(MethodSpec.methodBuilder("getViewStateProviders")
 					.addModifiers(Modifier.PUBLIC, Modifier.STATIC)
 					.returns(MAP_CLASS_TO_OBJECT_TYPE_NAME)
-					.addStatement("return viewStateProvider.getViewState()")
+					.addStatement("return sViewStateProviders")
 					.build());
 
 			classBuilder.addMethod(MethodSpec.methodBuilder("getPresenterBinders")
 					.addModifiers(Modifier.PUBLIC, Modifier.STATIC)
 					.returns(MAP_CLASS_TO_LIST_OF_OBJECT_TYPE_NAME)
-					.addStatement("return sViewStateProviders")
+					.addStatement("return sPresenterBinders")
 					.build());
 
 			classBuilder.addMethod(MethodSpec.methodBuilder("getStrategies")

--- a/moxy-compiler/src/test/java/com/arellomobile/mvp/compiler/CompilerTest.java
+++ b/moxy-compiler/src/test/java/com/arellomobile/mvp/compiler/CompilerTest.java
@@ -39,6 +39,13 @@ public abstract class CompilerTest {
 				.compile(sources);
 	}
 
+	protected Compilation compileLibSourcesWithProcessor(String moxyReflectorPackage, JavaFileObject... sources) {
+		return javac()
+				.withOptions("-implicit:none", "-AmoxyReflectorPackage=" + moxyReflectorPackage)
+				.withProcessors(new MvpCompiler())
+				.compile(sources);
+	}
+
 	/**
 	 * Asserts that all files from {@code exceptedGeneratedFiles} exists in {@code actualGeneratedFiles}
 	 * and have equivalent bytecode

--- a/moxy-compiler/src/test/java/com/arellomobile/mvp/compiler/MultiModulesTest.java
+++ b/moxy-compiler/src/test/java/com/arellomobile/mvp/compiler/MultiModulesTest.java
@@ -8,20 +8,53 @@ import javax.tools.JavaFileObject;
 
 import static com.google.testing.compile.CompilationSubject.assertThat;
 import static com.google.testing.compile.JavaFileObjects.forResource;
+import static com.google.testing.compile.JavaFileObjects.forSourceLines;
 
 public class MultiModulesTest extends CompilerTest {
 
 	@Test
-	public void testCompileLibraryModule() throws Exception {
-		JavaFileObject presenter = forResource("multimodules/lib1/Lib1Presenter.java");
-		JavaFileObject view = forResource("multimodules/lib1/Lib1View.java");
+	public void testLibraryModule() throws Exception {
+		JavaFileObject[] sources = {
+				forResource("multimodules/lib1/Lib1Presenter.java"),
+				forResource("multimodules/lib1/Lib1View.java")
+		};
 
-		Compilation compilation = compileLibSourcesWithProcessor("multimodules.lib1", presenter, view);
-		Compilation exceptedCompilation = compileSources(
+		JavaFileObject[] generatedSources = {
 				forResource("multimodules/lib1/Lib1Presenter$$ViewStateProvider.java"),
 				forResource("multimodules/lib1/Lib1View$$State.java"),
 				forResource("multimodules/lib1/MoxyReflector.java")
+		};
+
+		Compilation compilation = compileLibSourcesWithProcessor("multimodules.lib1", sources);
+		Compilation exceptedCompilation = compileSources(generatedSources);
+
+		assertThat(compilation).succeededWithoutWarnings();
+		assertExceptedFilesGenerated(compilation.generatedFiles(), exceptedCompilation.generatedFiles());
+	}
+
+	@Test
+	public void testRegisterMoxyReflectorPackages() throws Exception {
+		JavaFileObject someClientClass = forSourceLines("multimodules.app.App",
+				"package multimodules.app;",
+				"import com.arellomobile.mvp.RegisterMoxyReflectorPackages;",
+				"@RegisterMoxyReflectorPackages(\"multimodules.lib1\")",
+				"public class App {}"
 		);
+
+		JavaFileObject[] sources = new JavaFileObject[]{
+				forResource("multimodules/app/AppPresenter.java"),
+				forResource("multimodules/app/AppView.java"),
+				someClientClass
+		};
+
+		JavaFileObject[] exceptedSources = new JavaFileObject[]{
+				forResource("multimodules/app/AppPresenter$$ViewStateProvider.java"),
+				forResource("multimodules/app/AppView$$State.java"),
+				forResource("multimodules/app/MoxyReflector.java")
+		};
+
+		Compilation compilation = compileSourcesWithProcessor(sources);
+		Compilation exceptedCompilation = compileSources(exceptedSources);
 
 		assertThat(compilation).succeededWithoutWarnings();
 		assertExceptedFilesGenerated(compilation.generatedFiles(), exceptedCompilation.generatedFiles());

--- a/moxy-compiler/src/test/java/com/arellomobile/mvp/compiler/MultiModulesTest.java
+++ b/moxy-compiler/src/test/java/com/arellomobile/mvp/compiler/MultiModulesTest.java
@@ -1,0 +1,29 @@
+package com.arellomobile.mvp.compiler;
+
+import com.google.testing.compile.Compilation;
+
+import org.junit.Test;
+
+import javax.tools.JavaFileObject;
+
+import static com.google.testing.compile.CompilationSubject.assertThat;
+import static com.google.testing.compile.JavaFileObjects.forResource;
+
+public class MultiModulesTest extends CompilerTest {
+
+	@Test
+	public void testCompileLibraryModule() throws Exception {
+		JavaFileObject presenter = forResource("multimodules/lib1/Lib1Presenter.java");
+		JavaFileObject view = forResource("multimodules/lib1/Lib1View.java");
+
+		Compilation compilation = compileLibSourcesWithProcessor("multimodules.lib1", presenter, view);
+		Compilation exceptedCompilation = compileSources(
+				forResource("multimodules/lib1/Lib1Presenter$$ViewStateProvider.java"),
+				forResource("multimodules/lib1/Lib1View$$State.java"),
+				forResource("multimodules/lib1/MoxyReflector.java")
+		);
+
+		assertThat(compilation).succeededWithoutWarnings();
+		assertExceptedFilesGenerated(compilation.generatedFiles(), exceptedCompilation.generatedFiles());
+	}
+}

--- a/moxy-compiler/src/test/resources/multimodules/app/AppPresenter$$ViewStateProvider.java
+++ b/moxy-compiler/src/test/resources/multimodules/app/AppPresenter$$ViewStateProvider.java
@@ -1,0 +1,12 @@
+package multimodules.app;
+
+import com.arellomobile.mvp.MvpView;
+import com.arellomobile.mvp.ViewStateProvider;
+import com.arellomobile.mvp.viewstate.MvpViewState;
+
+public class AppPresenter$$ViewStateProvider extends ViewStateProvider {
+	@Override
+	public MvpViewState<? extends MvpView> getViewState() {
+		return new AppView$$State();
+	}
+}

--- a/moxy-compiler/src/test/resources/multimodules/app/AppPresenter.java
+++ b/moxy-compiler/src/test/resources/multimodules/app/AppPresenter.java
@@ -1,0 +1,8 @@
+package multimodules.app;
+
+import com.arellomobile.mvp.InjectViewState;
+import com.arellomobile.mvp.MvpPresenter;
+
+@InjectViewState
+public class AppPresenter extends MvpPresenter<AppView> {
+}

--- a/moxy-compiler/src/test/resources/multimodules/app/AppView$$State.java
+++ b/moxy-compiler/src/test/resources/multimodules/app/AppView$$State.java
@@ -1,0 +1,6 @@
+package multimodules.app;
+
+import com.arellomobile.mvp.viewstate.MvpViewState;
+
+public class AppView$$State extends MvpViewState<AppView> implements AppView {
+}

--- a/moxy-compiler/src/test/resources/multimodules/app/AppView.java
+++ b/moxy-compiler/src/test/resources/multimodules/app/AppView.java
@@ -1,0 +1,6 @@
+package multimodules.app;
+
+import com.arellomobile.mvp.MvpView;
+
+public interface AppView extends MvpView {
+}

--- a/moxy-compiler/src/test/resources/multimodules/app/MoxyReflector.java
+++ b/moxy-compiler/src/test/resources/multimodules/app/MoxyReflector.java
@@ -1,0 +1,47 @@
+package com.arellomobile.mvp;
+
+import java.lang.Class;
+import java.lang.Object;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import multimodules.app.AppPresenter;
+import multimodules.app.AppPresenter$$ViewStateProvider;
+
+public final class MoxyReflector {
+	private static Map<Class<?>, Object> sViewStateProviders;
+
+	private static Map<Class<?>, List<Object>> sPresenterBinders;
+
+	private static Map<Class<?>, Object> sStrategies;
+
+	static {
+		sViewStateProviders = new HashMap<>();
+		sViewStateProviders.put(AppPresenter.class, new AppPresenter$$ViewStateProvider());
+
+		sPresenterBinders = new HashMap<>();
+
+		sStrategies = new HashMap<>();
+
+		sViewStateProviders.putAll(multimodules.lib1.MoxyReflector.getViewStateProviders());
+		sPresenterBinders.putAll(multimodules.lib1.MoxyReflector.getPresenterBinders());
+		sStrategies.putAll(multimodules.lib1.MoxyReflector.getStrategies());
+	}
+
+	public static Object getViewState(Class<?> presenterClass) {
+		ViewStateProvider viewStateProvider = (ViewStateProvider) sViewStateProviders.get(presenterClass);
+		if (viewStateProvider == null) {
+			return null;
+		}
+
+		return viewStateProvider.getViewState();
+	}
+
+	public static List<Object> getPresenterBinders(Class<?> delegated) {
+		return sPresenterBinders.get(delegated);
+	}
+
+	public static Object getStrategy(Class<?> strategyClass) {
+		return sStrategies.get(strategyClass);
+	}
+}

--- a/moxy-compiler/src/test/resources/multimodules/lib1/Lib1Presenter$$ViewStateProvider.java
+++ b/moxy-compiler/src/test/resources/multimodules/lib1/Lib1Presenter$$ViewStateProvider.java
@@ -1,0 +1,12 @@
+package multimodules.lib1;
+
+import com.arellomobile.mvp.MvpView;
+import com.arellomobile.mvp.ViewStateProvider;
+import com.arellomobile.mvp.viewstate.MvpViewState;
+
+public class Lib1Presenter$$ViewStateProvider extends ViewStateProvider {
+	@Override
+	public MvpViewState<? extends MvpView> getViewState() {
+		return new Lib1View$$State();
+	}
+}

--- a/moxy-compiler/src/test/resources/multimodules/lib1/Lib1Presenter.java
+++ b/moxy-compiler/src/test/resources/multimodules/lib1/Lib1Presenter.java
@@ -1,0 +1,8 @@
+package multimodules.lib1;
+
+import com.arellomobile.mvp.InjectViewState;
+import com.arellomobile.mvp.MvpPresenter;
+
+@InjectViewState
+public class Lib1Presenter extends MvpPresenter<Lib1View> {
+}

--- a/moxy-compiler/src/test/resources/multimodules/lib1/Lib1View$$State.java
+++ b/moxy-compiler/src/test/resources/multimodules/lib1/Lib1View$$State.java
@@ -1,0 +1,6 @@
+package multimodules.lib1;
+
+import com.arellomobile.mvp.viewstate.MvpViewState;
+
+public class Lib1View$$State extends MvpViewState<Lib1View> implements Lib1View {
+}

--- a/moxy-compiler/src/test/resources/multimodules/lib1/Lib1View.java
+++ b/moxy-compiler/src/test/resources/multimodules/lib1/Lib1View.java
@@ -1,0 +1,6 @@
+package multimodules.lib1;
+
+import com.arellomobile.mvp.MvpView;
+
+public interface Lib1View extends MvpView {
+}

--- a/moxy-compiler/src/test/resources/multimodules/lib1/MoxyReflector.java
+++ b/moxy-compiler/src/test/resources/multimodules/lib1/MoxyReflector.java
@@ -1,0 +1,34 @@
+package multimodules.lib1;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public final class MoxyReflector {
+	private static Map<Class<?>, Object> sViewStateProviders;
+
+	private static Map<Class<?>, List<Object>> sPresenterBinders;
+
+	private static Map<Class<?>, Object> sStrategies;
+
+	static {
+		sViewStateProviders = new HashMap<>();
+		sViewStateProviders.put(Lib1Presenter.class, new Lib1Presenter$$ViewStateProvider());
+
+		sPresenterBinders = new HashMap<>();
+
+		sStrategies = new HashMap<>();
+	}
+
+	public static Map<Class<?>, Object> getViewStateProviders() {
+		return sViewStateProviders;
+	}
+
+	public static Map<Class<?>, List<Object>> getPresenterBinders() {
+		return sPresenterBinders;
+	}
+
+	public static Map<Class<?>, Object> getStrategies() {
+		return sStrategies;
+	}
+}

